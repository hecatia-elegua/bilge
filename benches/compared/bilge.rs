use bilge::{Bitsized, Number, FromBits, bitsize, u4, u12, u20, u5, u2, DebugBits};

#[inline(never)]
pub fn bilge(input: (u32, u32, u64, u16)) {
    let mut lpi = GicRedistributorLpi {
        control: RedistributorControl::from(input.0),
        implementer_identification: RedistributorImplementerIdentification::from(input.1),
        redistributor_type: RedistributorType::from(input.2),
    };
    assert_eq!(lpi.control.value, input.0);
    assert_eq!(lpi.implementer_identification.value, input.1);
    assert_eq!(lpi.redistributor_type.value, input.2);
    
    assert!(lpi.control.clear_enable_supported());
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(2054));
    lpi.implementer_identification.set_implementer_jep106(u12::new(input.3));
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(input.3));
    assert_eq!(lpi.redistributor_type.processor_number(), 63872);
}

#[derive(Debug)]
pub struct GicRedistributorLpi {
    control: RedistributorControl,
    implementer_identification: RedistributorImplementerIdentification,
    redistributor_type: RedistributorType,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorControl {
    //ro
    upstream_write_pending: bool,
    //zero
    reserved: u4,
    //or reserved
        disable_processor_selection_for_group_1_secure_interrupts: bool,
        disable_processor_selection_for_group_1_non_secure_interrupts: bool,
        disable_processor_selection_for_group_0_interrupts: bool,
    //zero
    reserved: u20,
    register_write_pending: bool,
    lpi_invalidate_registers_supported: bool,
    clear_enable_supported: bool,
    enable_lpis: bool,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorImplementerIdentification {
    //ro
    product_id: u8,
    //zero
    reserved: u4,
    //ro
    variant: u4,
    //ro
    revision: u4,
    implementer_jep106: u12,
}

#[bitsize(64)]
#[derive(DebugBits, FromBits)]
struct RedistributorType {
    affinity_value: u32,
    ppi_num: u5,
    virtual_sgi_supported: bool,
    common_lpi_affinity: u2, //TODO: enums everywhere
    processor_number: u16,
    resident_vpe_id: bool,
    mpam_supported: bool,
    control_cpgs_supported: bool,
    is_last: bool,
    direct_lpi_supported: bool,
    dirty: bool,
    virtual_lpi_supported: bool,
    physical_lpi_supported: bool,
}
