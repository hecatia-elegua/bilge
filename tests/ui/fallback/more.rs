use bilge::prelude::*;

#[bitsize(15)]
#[derive(FromBits)]
enum Testing {
    A,
    D,
    H,
    #[fallback]
    Dee { fallback: u15 },
}

#[bitsize(15)]
#[derive(FromBits)]
enum Windows {
    I,
    N,
    #[fallback]
    Tel(u8, u7),
}

#[bitsize(15)]
#[derive(FromBits)]
enum Alt {
    F,
    #[fallback]
    Four(Option<u8>),
}

#[bitsize(15)]
#[derive(FromBits)]
enum James {
    Atomic,
    #[fallback]
    Habits(u100),
}

#[bitsize(15)]
#[derive(FromBits)]
struct Dum {
    #[fallback]
    my: u15,
}

fn main() {}
