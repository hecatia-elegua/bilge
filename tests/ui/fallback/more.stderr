error: `#[fallback]` does not support variants with named fields
  --> tests/ui/fallback/more.rs:9:5
   |
9  | /     #[fallback]
10 | |     Dee { fallback: u15 },
   | |_________________________^
   |
   = help: use a tuple variant or remove this `#[fallback]`

error: fallback variant must have exactly one field
  --> tests/ui/fallback/more.rs:18:5
   |
18 | /     #[fallback]
19 | |     Tel(u8, u7),
   | |_______________^
   |
   = help: use only one field or change to a unit variant

error: `#[fallback]` only supports arbitrary_int or bool types
  --> tests/ui/fallback/more.rs:27:9
   |
27 |     Four(Option<u8>),
   |         ^^^^^^^^^^^^

error: bitsize of fallback field (100) does not match bitsize of enum (15)
  --> tests/ui/fallback/more.rs:35:11
   |
35 |     Habits(u100),
   |           ^^^^^^

error: `#[fallback]` is only applicable to enums
  --> tests/ui/fallback/more.rs:39:10
   |
39 | #[derive(FromBits)]
   |          ^^^^^^^^
   |
   = help: remove all `#[fallback]` from this struct
   = note: this error originates in the derive macro `FromBits` (in Nightly builds, run with -Z macro-backtrace for more info)
